src/lint/rules/algorithm-step-numbering.ts

```diff
@@ -0,0 +1,50 @@
+import type { Node as EcmarkdownNode, Observer } from 'ecmarkdown';
+import type { LintingError } from '../algorithm-error-reporter-type';
+
+const ruleId = 'algorithm-step-numbering';
+
+/*
+Checks that step numbers are all `1`, with the exception of top-level lists whose first item is not `1`.
+*/
+export default function (
+  report: (e: LintingError) => void,
+  node: Element,
+  algorithmSource: string
+): Observer {
+  const nodeType = node.tagName;
+  let depth = -1;
+  let topLevelIsNotOne = false;
+  return {
+    enter(node: EcmarkdownNode) {
+      if (node.name === 'ol') {
+        ++depth;
+        if (depth === 0) {
+          topLevelIsNotOne = node.start !== 1;
+        }
+      } else if (node.name === 'ordered-list-item') {
+        if (depth === 0 && topLevelIsNotOne) {
+          return;
+        }
+        let itemSource = algorithmSource.slice(
+          node.location!.start.offset,
+          node.location!.end.offset
+        );
+        let match = itemSource.match(/^([ \t]*)([^.]+\.) /)!;
```

Comment:
```suggestion
        let match = itemSource.match(/^(\s*)(\d+\.) /)!;
```

They always start with a number, right? Also, is it possible for the match to fail? Should we account for that?

Reply:
They do always start with a number, and it is not possible for the match to fail.